File: src/main/java/net/darkmorford/jas/block/BlockSnad.java
package net.darkmorford.jas.block;

import net.darkmorford.jas.JustAnotherSnad;
import net.darkmorford.jas.configuration.ConfigurationData;
import net.minecraft.block.*;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.IBlockState;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraftforge.common.EnumPlantType;
import net.minecraftforge.common.IPlantable;
import org.apache.logging.log4j.Level;

import java.util.Random;

import static net.darkmorford.jas.JustAnotherSnad.MODID;

public class BlockSnad extends BlockSand {

    public BlockSnad() {
        super();
        setCreativeTab(CreativeTabs.MISC);
        setHardness(0.5f);
        setSoundType(SoundType.SAND);
        setTickRandomly(true);
        setUnlocalizedName("snad");
        setRegistryName(MODID, "snad");
    }

    @Override
    public void randomTick(World world, BlockPos position, IBlockState state, Random random) {
        super.randomTick(world, position, state, random);

        Block blockAbove = world.getBlockState(position.up()).getBlock();
        if (blockAbove instanceof BlockReed || blockAbove instanceof BlockCactus) {
            checkColumnOfPlant(world, position, random, blockAbove);
        } else if (blockAbove instanceof IPlantable) {
            blockAbove.randomTick(world, position.up(), world.getBlockState(position.up()), random);
        }
    }

    private void checkColumnOfPlant(World world, BlockPos originalPosition, Random random, Block blockAbove) {
        Block nextPlantBlock;
        BlockPos nextPosition = originalPosition.up();
        while ((nextPlantBlock = world.getBlockState(nextPosition).getBlock()).getClass().equals(blockAbove.getClass())) {
            growthLoop(world, originalPosition, random, (IPlantable) blockAbove, nextPosition, nextPlantBlock);
            nextPosition = nextPosition.up();
        }
    }

    private void growthLoop(World world, BlockPos orignalPosition, Random random, IPlantable blockAbove, BlockPos nextPosition, Block nextPlantBlock) {
        for (int growthAttempts = 0; growthAttempts < ConfigurationData.SNAD_SPEED_INCREASE_VALUE; growthAttempts++) {
            if (growthAttempts == 0 || canSustainPlant(world.getBlockState(orignalPosition), world, orignalPosition, null, blockAbove)) {
                nextPlantBlock.randomTick(world, nextPosition, world.getBlockState(nextPosition), random);
            }
        }
    }

    @Override
    public boolean canSustainPlant(IBlockState state, IBlockAccess world, BlockPos position, EnumFacing facing, IPlantable plantable) {
        BlockPos plantPosition = new BlockPos(position.getX(), position.getY() + 1, position.getZ());
        EnumPlantType plantType = plantable.getPlantType(world, plantPosition);
        switch (plantType) {
            case Desert:
                return true;
            case Beach:
                return ((world.getBlockState(new BlockPos(position.getX() - 1, position.getY(), position.getZ())).getMaterial() == Material.WATER) ||
                    (world.getBlockState(new BlockPos(position.getX() + 1, position.getY(), position.getZ())).getMaterial() == Material.WATER) ||
                    (world.getBlockState(new BlockPos(position.getX(), position.getY(), position.getZ() + 1)).getMaterial() == Material.WATER) ||
                    (world.getBlockState(new BlockPos(position.getX(), position.getY(), position.getZ() - 1)).getMaterial() == Material.WATER));

            case Water:
                return (world.getBlockState(position).getMaterial() == Material.WATER) && (world.getBlockState(position) == getDefaultState());
            default: {
                return false;
            }
        }
    }
}

File: src/main/kotlin/net/darkmorford/jas/block/BlockSnad.kt
package net.darkmorford.jas.block

import net.darkmorford.jas.JustAnotherSnad
import net.darkmorford.jas.item.IMetaBlockSnad
import net.minecraft.block.SoundType
import net.minecraft.block.material.Material
import net.minecraft.block.properties.PropertyEnum
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraftforge.common.EnumPlantType
import net.minecraftforge.common.IPlantable

class BlockSnad : AbsBlockSnad(), IMetaBlockSnad {

	override fun getSpecialName(stack: ItemStack): String {
		return if (stack.itemDamage == 0) "default" else "red"
	}

	override fun createBlockState(): BlockStateContainer {
		return BlockStateContainer(this, VARIANT)
	}

	override fun getStateFromMeta(meta: Int): IBlockState {
		return defaultState.withProperty(VARIANT, EnumType.byMetadata(meta))
	}

	override fun getMetaFromState(state: IBlockState): Int {
		return state.getValue(VARIANT).metadata
	}

	override fun canSustainPlant(state: IBlockState, world: IBlockAccess, position: BlockPos, facing : EnumFacing, plantable: IPlantable): Boolean {
		val plantPosition = BlockPos(position.x, position.y + 1, position.z)
		val plantType = plantable.getPlantType(world, plantPosition)

		return when (plantType) {
			EnumPlantType.Desert -> true
			EnumPlantType.Water -> (world.getBlockState(position).material  == Material.WATER) && (world.getBlockState(position) == defaultState)
			EnumPlantType.Beach -> (
					(world.getBlockState(BlockPos(position.x - 1, position.y, position.z)).material == Material.WATER) ||
							(world.getBlockState(BlockPos(position.x + 1, position.y, position.z)).material == Material.WATER) ||
							(world.getBlockState(BlockPos(position.x, position.y, position.z + 1)).material == Material.WATER) ||
							(world.getBlockState(BlockPos(position.x, position.y, position.z - 1)).material == Material.WATER)
					)
			else -> false
		}
	}

	init {
		setCreativeTab(CreativeTabs.MISC)
		setHardness(0.5f)
		soundType = SoundType.SAND
		tickRandomly = true

		unlocalizedName = "snad"
		defaultState = blockState.baseState.withProperty(VARIANT, EnumType.SNAD)
		setRegistryName(JustAnotherSnad.MODID, "snad")
	}

	companion object {
		private val snadClass = EnumType::class
		private val snadJavaClass = snadClass.java
		private val VARIANT: PropertyEnum<AbsBlockSnad.EnumType> = PropertyEnum.create("variant", snadJavaClass)
	}


}
